File: src/migrations.ts
import {getKingdom, saveKingdom} from './kingdom/storage';
import {getNumberSetting, setSetting} from './settings';
import {getCamping, saveCamping} from './camping/storage';
import {isFirstGm} from './utils';
import {Migration} from './migrations/migration';
import {Migration2} from './migrations/migration2';
import {Migration3} from './migrations/migration3';
import {Migration4} from './migrations/migration4';
import {Migration5} from './migrations/migration5';
import {Migration6} from './migrations/migration6';
import {Migration7} from './migrations/migration7';
import {Migration8} from './migrations/migration8';
import {Migration9} from './migrations/migration9';


const migrations: Migration[] = [
    new Migration2(),
    new Migration3(),
    new Migration4(),
    new Migration5(),
    new Migration6(),
    new Migration7(),
    new Migration8(),
    new Migration9(),
];

interface BackupParams {
    game: Game;
    kingdomActor: Actor | undefined | null;
    campingActor: Actor | null | undefined;
    currentVersion: number;
}

async function createBackups(params: BackupParams): Promise<void> {
    const backup = {
        version: params.currentVersion,
        camping: params.campingActor ? getCamping(params.campingActor) : null,
        kingdom: params.kingdomActor ? getKingdom(params.kingdomActor) : null,
    };
    await setSetting(params.game, 'latestMigrationBackup', JSON.stringify(backup));
}

export async function migrate(game: Game, kingdomActor: Actor | undefined | null, campingActor: Actor | null | undefined): Promise<void> {
    const currentVersion = getNumberSetting(game, 'schemaVersion');
    const latestMigration = Math.max(1, ...migrations.map(m => m.version));
    if (isFirstGm(game) && currentVersion < latestMigration) {
        ui.notifications?.info('Running Kingmaker Tools Migrations, please do not close the window');
        await createBackups({game, kingdomActor, campingActor, currentVersion});

        const migrationsToRun = migrations.filter(m => m.version > currentVersion);

        if (kingdomActor) {
            const kingdom = getKingdom(kingdomActor);
            for (const migration of migrationsToRun) {
                await migration.migrateKingdom(game, kingdom);
            }
            await saveKingdom(kingdomActor, kingdom);
        }

        if (campingActor) {
            const camping = getCamping(campingActor);
            for (const migration of migrationsToRun) {
                await migration.migrateCamping(game, camping);
            }
            await saveCamping(game, campingActor, camping);
        }

        for (const migration of migrationsToRun) {
            await migration.migrateOther(game);
        }

        await setSetting(game, 'schemaVersion', latestMigration);
        ui.notifications?.info('Successfully migrated Kingmaker tools');
    }
}

File: src/jsMain/kotlin/at/posselt/pfrpg2e/migrations/Migrations.kt
package at.posselt.pfrpg2e.migrations

import at.posselt.pfrpg2e.camping.CampingActor
import at.posselt.pfrpg2e.camping.getCamping
import at.posselt.pfrpg2e.camping.getCampingActors
import at.posselt.pfrpg2e.camping.setCamping
import at.posselt.pfrpg2e.kingdom.KingdomActor
import at.posselt.pfrpg2e.kingdom.getKingdom
import at.posselt.pfrpg2e.kingdom.getKingdomActors
import at.posselt.pfrpg2e.kingdom.setKingdom
import at.posselt.pfrpg2e.migrations.migrations.Migration17
import at.posselt.pfrpg2e.settings.pfrpg2eKingdomCampingWeather
import at.posselt.pfrpg2e.utils.isFirstGM
import at.posselt.pfrpg2e.utils.openJournal
import at.posselt.pfrpg2e.utils.t
import at.posselt.pfrpg2e.utils.toRecord
import com.foundryvtt.core.Game
import com.foundryvtt.core.ui
import js.objects.recordOf

private suspend fun createBackups(
    game: Game,
    kingdomActors: List<KingdomActor>,
    campingActors: List<CampingActor>,
    currentVersion: Int
) {
    val backup = recordOf(
        "version" to currentVersion,
        "camping" to campingActors.map {
            it.id!! to it.getCamping()
        }.toRecord(),
        "kingdoms" to kingdomActors.map {
            it.id!! to it.getKingdom()
        }.toRecord()
    )
    game.settings.pfrpg2eKingdomCampingWeather.setLatestMigrationBackup(
        JSON.stringify(backup)
    )
}

private val migrations = listOf(
    Migration17(),
)

private val latestMigrationVersion = migrations.maxOfOrNull { it.version }!!

suspend fun Game.migratePfrpg2eKingdomCampingWeather() {
    val currentVersion = settings.pfrpg2eKingdomCampingWeather.getSchemaVersion()
        .takeIf { it != 0 }
        ?: latestMigrationVersion
    console.log("${t("moduleName")}: ${t("migrations.upgradingFromTo", recordOf("fromVersion" to currentVersion, "toVersion" to latestMigrationVersion))}")
    if (currentVersion < 16) {
        ui.notifications.error(
            "${t("moduleName")}: ${t("migrations.unsupportedVersions", recordOf("version" to "4.8.2 (FoundryVTT V12)"))}"
        )
    } else if (isFirstGM() && currentVersion < latestMigrationVersion) {
        try {
            migrateFrom(currentVersion)
        } catch (e: Throwable) {
            console.log(e)
            ui.notifications.error(t("migrations.failed", recordOf("version" to currentVersion)))
            throw e
        }
    }
}

private suspend fun Game.migrateFrom(currentVersion: Int) {
    ui.notifications.info("${t("moduleName")}: ${t("migrations.doNotClose")}")
    // create backups
    val kingdomActors = getKingdomActors()
    val campingActors = getCampingActors()
    createBackups(this, kingdomActors, campingActors, currentVersion)

    val migrationsToRun = migrations.filter { it.version > currentVersion }

    migrationsToRun
        .forEach { migration ->
            ui.notifications.info(
                "${t("moduleName")}: ${
                    t(
                        "migrations.runningMigration",
                        recordOf("version" to migration.version)
                    )
                }"
            )
            campingActors.forEach { actor ->
                actor.getCamping()?.let { camping ->
                    migration.migrateCamping(this, camping)
                    actor.setCamping(camping)
                }
            }

            kingdomActors.forEach { actor ->
                actor.getKingdom()?.let { kingdom ->
                    migration.migrateKingdom(this, kingdom)
                    actor.setKingdom(kingdom)
                }
            }

            migration.migrateOther(this)
        }

    settings.pfrpg2eKingdomCampingWeather.setSchemaVersion(latestMigrationVersion)
    ui.notifications.info("${t("moduleName")}: ${t("migrations.successful")}")

    if (migrationsToRun.any { it.showUpgradingNotices }) {
        openJournal("Compendium.pf2e-kingmaker-tools.kingmaker-tools-journals.JournalEntry.wz1mIWMxDJVsMIUd")
    }
}

